packages/report-server/src/reportBuilder/transform/functions/groupRows/groupRows.ts

```diff
@@ -0,0 +1,130 @@
+import { functions } from '../../../functions';
+
+type GroupRowsParams = {
+  createGroupKey: (row: Row) => string;
+  getMergeStrategy: (field: string) => keyof typeof mergeStrategies;
+  where: (parser: TransformParser) => boolean;
+};
+
+const paramsValidator = yup.object().shape({
+  by: yup.lazy((value: unknown) => {
+    if (typeof value === 'string' || value === undefined) {
+      return yup.string();
+    }
+
+    if (Array.isArray(value)) {
+      return yup.array().of(yup.string().required());
+    }
+
+    throw new yup.ValidationError('by must be either a single column, or an array of columns');
+  }),
+  mergeUsing: yup.lazy((value: unknown) => {
+    const mergeStrategyNameValidator = yup
+      .mixed<keyof typeof mergeStrategies>()
+      .oneOf(Object.keys(mergeStrategies) as (keyof typeof mergeStrategies)[])
+      .required();
+    if (typeof value === 'string') {
+      return mergeStrategyNameValidator;
+    }
+
+    if (typeof value === 'object' && value !== null) {
+      const mergeStrategyMapValidator = Object.fromEntries(
+        Object.entries(value).map(([columnName]) => [columnName, mergeStrategyNameValidator]),
+      );
+      return yup.object().shape(mergeStrategyMapValidator);
+    }
+
+    throw new yup.ValidationError(
+      'mergeUsing must be either a single merge strategy, or a mapping between columns and merge strategies',
+    );
+  }),
+  where: yup.string(),
+});
+
+/**
+ * A group represents a key of fields, to all the values each row in that group had for the respective field
+ *
+ * eg.
+ * [{orgUnit: TO, BCD1: 4 }, {orgUnit: TO, BCD1: 7 }], groupBy orgUnit => group: { BCD1: [4, 7] }
+ */
+type Group = {
+  [fieldKey: string]: FieldValue[];
+};
+
+const buildGroups = (rows: Row[], params: GroupRowsParams) => {
+  const groupsByKey: Record<string, Group> = {};
+  const parser = new TransformParser(rows, functions);
+  const ungroupedRows: Row[] = []; // Rows that don't match the 'where' clause are left ungrouped
+
+  rows.forEach((row: Row) => {
+    if (!params.where(parser)) {
+      ungroupedRows.push(row);
+      parser.next();
+      return;
+    }
+    const groupKey = params.createGroupKey(row);
+    groupsByKey[groupKey] = merge(groupsByKey[groupKey] || {}, row);
+    parser.next();
+  });
+
+  return { groups: Object.values(groupsByKey), ungroupedRows };
+};
+
+const merge = (previousRow: Group, newRow: Row): Group => {
+  const mergedRow = { ...previousRow };
+  Object.keys(newRow).forEach((field: string) => {
+    mergedRow[field] = [...(mergedRow[field] || []), newRow[field]];
+  });
+  return mergedRow;
+};
+
+const mergeGroups = (groups: Group[], params: GroupRowsParams): Row[] => {
+  return groups.map(group => {
+    const mergedRow: Row = {};
+    Object.entries(group).forEach(([fieldKey, fieldValue]) => {
+      const mergeStrategy = params.getMergeStrategy(fieldKey);
+      const mergedValue = mergeStrategies[mergeStrategy](fieldValue);
```

Comment:
Good call! Btw I've been considering doing a bulk-rename of `FieldValue` to `CellValue` and instead of `field` using the term `column`, eg.

```
type Row = {
  [column: string]: CellValue;
}
```
?

Probably not as part of this PR, but curious to know if you like the abstraction?

Reply:
Yep I prefer that :+1: